staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go

```diff
@@ -96,33 +96,39 @@ func BuildSwagger(crd *apiextensions.CustomResourceDefinition, version string) (
 
 	routes := make([]*restful.RouteBuilder, 0)
 	root := fmt.Sprintf("/apis/%s/%s/%s", b.group, b.version, b.plural)
+
+	var params []*restful.Parameter
+
 	if b.namespaced {
-		routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList).
-			Operation("list"+b.kind+"ForAllNamespaces"))
+		namespaceParam := b.ws.PathParameter("namespace", "object name and auth scope, such as for teams and projects").DataType("string")
+		params = append(params, namespaceParam)
+		routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList, params).Operation("list"+b.kind+"ForAllNamespaces"))
 		root = fmt.Sprintf("/apis/%s/%s/namespaces/{namespace}/%s", b.group, b.version, b.plural)
 	}
-	routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList))
-	routes = append(routes, b.buildRoute(root, "", "POST", "create", sample).Reads(sample))
-	routes = append(routes, b.buildRoute(root, "", "DELETE", "deletecollection", status))
+	routes = append(routes, b.buildRoute(root, "", "GET", "list", sampleList, params))
+	routes = append(routes, b.buildRoute(root, "", "POST", "create", sample, params).Reads(sample))
+	routes = append(routes, b.buildRoute(root, "", "DELETE", "deletecollection", status, params))
```

Comment:
Thanks for catching and fixing this!

/lgtm

nit: I was thinking of two alternatives (to save some code; not sure if it's more readable or not):

1. avoid passing `params` to `buildRoute`
```
func (b *builder) buildRoute(root, path, action, verb string, sample interface{}) *restful.RouteBuilder {
    ...
+    // add path parameters to route, kube-openapi will aggregate the path parameters later
+    if strings.Contain(root, `{namespace}`) {
+        route.Param(b.ws.PathParameter("namespace", "object name and auth scope, such as for teams and projects").DataType("string"))
+    }
+    if strings.Contain(path, `{name}`) {
+        route.Param(b.ws.PathParameter("name", "name of the "+b.kind).DataType("string"))
+    }
}
```

2. 

one thing I noticed is that in kube-openapi we [aggregate parameters appears in all ops into common path parameters](https://github.com/kubernetes/kubernetes/blob/768acf600943442d517d16fdaaa786032099c995/vendor/k8s.io/kube-openapi/pkg/builder/openapi.go#L202-L220). You could see "namespace" and "name" fall under path parameter instead of op parameter in [the final spec](https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/swagger.json)

although this change technically behaves same as the [native API installer](https://github.com/kubernetes/kubernetes/blob/768acf600943442d517d16fdaaa786032099c995/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L428-L437) (it also add "namespace" and "name" as op parameters first), and I believe `openapibuilder.BuildOpenAPISpec` will do the same aggregation to the CRD openapi spec, I was imagining if we could add "namespace" and "name" as path parameters directly and retrieve them [here](https://github.com/kubernetes/kubernetes/blob/768acf600943442d517d16fdaaa786032099c995/vendor/k8s.io/kube-openapi/pkg/builder/openapi.go#L189) instead of going through `findCommonParameters`
```
if b.namespaced {
    ...
+    // add path parameters
+    b.ws.Param(b.ws.PathParameter("namespace", "object name and auth scope, such as for teams and projects").DataType("string"))
}
+ b.ws.Param(b.ws.PathParameter("name", "name of the "+b.kind).DataType("string"))
```
It turns out that we need two `ws`s to do this because we have two patterns for path parameters (w/ and w/o "name"). Still achievable, but kind of weird to read

Reply:
> 1. avoid passing params to buildRoute

mmm i didnt see much difference between the two. but conditioning by matching substrings from the path builds an implicit dependency on that `{}` format. even tho we dont see any plans to replace restful w/ another http framework. 

> one thing I noticed is that in kube-openapi we aggregate parameters appears in all ops into common path parameters.

yes as it's required by openapi spec doc structure. for now, `name` and `namespace` are actually fake path parameter in restful framework. we not reading from it instead we extracting them [from a filter](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go) so it's probably fine whether we misses that path parameter or not. the main reason we use restful is the openapi, so it would be better keep the integrity of openapi-related settings in the restful. i.e. let the builder pkg loads complete openapi information from restful instead of transforming it. wdyt?